flytekit/extras/tensorflow/record.py

```diff
@@ -0,0 +1,198 @@
+        local_path = ctx.file_access.get_random_local_path()
+        ctx.file_access.get_data(uri, local_path, is_multipart=False)
+        filenames = [local_path]
+        return tf.data.TFRecordDataset(
+            filenames=filenames,
+            compression_type=metadata.compression_type,
+            buffer_size=metadata.buffer_size,
+            num_parallel_reads=metadata.num_parallel_reads,
+            name=metadata.name,
+        )
+
+    def guess_python_type(self, literal_type: LiteralType) -> Type[TFRecordFile]:
+        if (
+            literal_type.blob is not None
+            and literal_type.blob.dimensionality == _core_types.BlobType.BlobDimensionality.SINGLE
+            and literal_type.blob.format == self.TENSORFLOW_FORMAT
+        ):
+            return TFRecordFile
+
+        raise ValueError(f"Transformer {self} cannot reverse {literal_type}")
+
+
+class TensorFlowRecordsDirTransformer(TypeTransformer[TFRecordsDirectory]):
+    """
+    TypeTransformer that supports serialising and deserialising to and from TFRecord directory.
+    https://www.tensorflow.org/tutorials/load_data/tfrecord
+    """
+
+    TENSORFLOW_FORMAT = "TensorFlowRecord"
+
+    def __init__(self):
+        super().__init__(name="TensorFlow Record Directory", t=TFRecordsDirectory)
+
+    def get_literal_type(self, t: Type[TFRecordsDirectory]) -> LiteralType:
+        return LiteralType(
+            blob=_core_types.BlobType(
+                format=self.TENSORFLOW_FORMAT,
+                dimensionality=_core_types.BlobType.BlobDimensionality.MULTIPART,
+            )
+        )
+
+    def to_literal(
+        self,
+        ctx: FlyteContext,
+        python_val: TFRecordsDirectory,
+        python_type: Type[TFRecordsDirectory],
+        expected: LiteralType,
+    ) -> Literal:
+        meta = BlobMetadata(
+            type=_core_types.BlobType(
+                format=self.TENSORFLOW_FORMAT,
+                dimensionality=_core_types.BlobType.BlobDimensionality.MULTIPART,
+            )
+        )
+        local_dir = ctx.file_access.get_random_local_directory()
+        remote_path = ctx.file_access.get_random_remote_directory()
+        for i, val in enumerate(python_val):
+            local_path = f"{local_dir}/part_{i}.tfrecord"
+            with tf.io.TFRecordWriter(local_path) as writer:
+                writer.write(val.SerializeToString())
+        ctx.file_access.upload_directory(local_dir, remote_path)
+        return Literal(scalar=Scalar(blob=Blob(metadata=meta, uri=remote_path)))
+
+    def to_python_value(
+        self, ctx: FlyteContext, lv: Literal, expected_python_type: Type[TFRecordsDirectory]
+    ) -> TFRecordDatasetV2:
+
+        uri, metadata = extract_metadata_and_uri(lv, expected_python_type)
+        local_dir = ctx.file_access.get_random_local_directory()
+        files = os.scandir(uri)
+        filenames = [os.path.join(local_dir, f.name) for f in files]
```

Comment:
I think we need to get the file names from the local directory, not the remote path. In local case, it works, but when run on Flyte backend, it'll be a remote uri.